File: data-structures/php/src/LinkedList/SinglyLinkedList/SinglyLinkedList.php
<?php

declare(strict_types=1);

namespace DataStructure\LinkedList\SinglyLinkedList;

use DataStructure\LinkedList\SinglyLinkedList\Node;
use Iterator;

class SinglyLinkedList implements Iterator{
    public ?Node $head = null;
    public ?Node $tail = null;
    public int $length = 0;

    // For Iterator implementation 
    // Not specifically related to the linked list
    private ?Node $currentNode;
    private int $currentKey = 0;


    public function __construct() {
    }

    // Push node to the end of the list
    public function push(string $data): bool {
        $node = new Node($data);

        // If the list is empty
        // then make this node as head and tail
        if ($this->head === null) {
            $this->head = &$node;
        } else {
            // set the current tail's next value to new node
            $this->tail->next = &$node;
        }

        // Set the tail to this new node
        $this->tail = &$node;

        // Increase the length by 1
        $this->length++;

        return true;
    }

    // Pop the last item
    public function pop(): ?Node {
        // If there is no head
        // then the list does not exist
        if ($this->head === null) {
            return null;
        }

        $current = $this->head;
        $newTail = $current;

        // Traverse through the list
        // and get the second last item as newTail
        while ($current->next !== null) {
            $newTail = $current;
            $current = $current->next;
        }

        // Assign the newTail to the tail
        $this->tail = &$newTail;

        // Delete the next pointer of the tail
        $this->tail->next = null;

        $this->length--;

        // Reset head and tail if it was the last item
        if ($this->length == 0) {
            $this->head = null;
            $this->tail = null;
        }

        // Return the popped item
        return $current;
    }

    // Traverse the list
    public function traverse(): void {
        $current = $this->head;

        while ($current !== null) {
            echo $current->data . PHP_EOL;

            // Change current to the next node
            $current = $current->next;
        }
    }

    // Shift head
    // and make the second item as head
    // also return the first item
    public function shift(): ?Node {
        // If there is no head then return null
        if ($this->head === null) {
            return null;
        }

        $oldHead = $this->head;

        // Set the second item as head
        $this->head = $oldHead->next;

        // Reduce the length by one
        $this->length--;

        if ($this->length === 0) {
            $this->head = null;
            $this->tail = null;
        }

        // Return the head item
        return $oldHead;
    }

    // Unshift head
    // Create new node and set that as head
    public function unshift($value): bool {
        // Create new Node from value
        $newHead = new Node($value);

        if ($this->head === null) {
            $this->tail = &$newHead;
        } else {
            $newHead->next = $this->head;
        }

        // Set the new node as head
        $this->head = $newHead;

        // Increase length by one
        $this->length++;

        return true;
    }

    // Get node by index(sequence numbers)
    // 0 based index (index starts from 0)
    public function get(int $index): ?Node {
        // Check if the index is valid
        if ($index < 0 || $index >= $this->length) {
            return null;
        }

        $selectedNode = $this->head;

        for ($i = 1; $i <= $index; $i++) {
            $selectedNode = $selectedNode->next;
        }

        return $selectedNode;
    }

    // Set data of a specific node at specific index
    // 0 based index(index starts from 0)
    public function set(int $index, string $data): bool {
        $nodeAtIndex = $this->get($index);

        if ($nodeAtIndex) {
            $nodeAtIndex->data = $data;
            return true;
        }

        return false;
    }

    // Search the list for specific data
    public function search(string $data): ?int {
        $current = $this->head;
        $index = 0;

        while ($current) {
            if ($current->data === $data) {
                return $index;
            }

            $current = $current->next;
            $index++;
        }

        return null;
    }

    // Insert new node at a specific index
    // 0 based index(index starts at 0)
    public function insert(int $index, string $data): bool {
        if ($index < 0 || $index > $this->length) {
            return false;
        }

        if ($index === $this->length) {
            $this->push($data);
            return true;
        }

        if ($index === 0) {
            $this->unshift($data);
            return true;
        }

        $newNode = new Node($data);
        $prevNode = $this->get($index - 1);
        $newNode->next = $prevNode->next;
        $prevNode->next = $newNode;

        // Increase length
        $this->length++;

        return true;
    }

    // Remove item at index
    public function remove(int $index): ?Node {
        // Return false if index is out of range
        if ($index < 0 || $index >= $this->length) {
            return null;
        }

        if ($index === 0) {
            return $this->shift();
        }

        if ($index === ($this->length - 1)) {
            return $this->pop();
        }

        $prevNode = $this->get($index - 1);
        $removedNode = $prevNode->next;
        $prevNode->next = $prevNode->next->next;

        $this->length--;

        return $removedNode;
    }

    // Reverse a linked list
    public function reverse(): self {
        $currentNode = $this->head;
        unset($this->tail);
        $prevNode = null;
        $nextNode = null;

        while ($currentNode !== null) {
            $nextNode = $currentNode->next;
            $currentNode->next = $prevNode;

            $prevNode = $currentNode;
            $currentNode = $nextNode;
        }

        $this->tail = $this->head;
        $this->head = $prevNode;
        
        return $this;
    }

    // Part of Iterator interface implementation
    // Get current node
    public function current(): string {
        return $this->currentNode->data;
    }

    // Part of Iterator interface implementation
    // Get current key
    public function key(): int {
        return $this->currentKey;
    }

    // Part of Iterator interface implementation
    // Move to the next item
    public function next(): void {
        $this->currentNode = $this->currentNode->next;
        $this->currentKey++;
    }

    // Part of Iterator interface implementation
    // Rewind the iterator
    public function rewind(): void {
        $this->currentKey = 0;
        $this->currentNode = $this->head;
    }

    // Part of Iterator interface implementation
    // Check is valid
    public function valid(): bool {
        return $this->currentNode !== null;
    }
    
}

File: data-structures/python/linked_list/singly_linked_list/singly_linked_list.py
from .node import Node

class SinglyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0
        
    # Push item to the end
    def push(self, data):
        # Create a new node
        node = Node(data)
        
        # If there is no head, that means the list is empty
        # Then make this new node as head
        if not self.head:
            self.head = node
        else:
            # Set the new node as the next of the current tail
            self.tail.next = node
            
        # Set the new node as tail
        self.tail = node
        
        # Increase length by 1
        self.length += 1
        
    # Pop item from the end
    def pop(self):
        # Return None if the list is empty
        if not self.head:
            return None
        
        current = self.head
        new_tail = current
        
        # Traverse through the list 
        # and get the second last item as new_tail
        while current.next:
            new_tail = current
            current = current.next
            
        # Set the new tail
        self.tail = new_tail
        
        # Set next of the new tail as None
        self.tail.next = None
        
        # Decrease the length by 1
        self.length -= 1
        
        # If there is no item left
        # then set the head and tail both as None
        if self.length == 0:
            self.head = None
            self.tail = None
            
        # Return the node in the current variable
        return current
    
    # Shift the head
    # and maek the second item as the head
    # Return the first item as result of shift
    def shift(self):
        # If the current head is None,
        # then return None
        if self.head is None:
            return None
        
        old_head = self.head
        
        # Set the next node as head
        self.head = old_head.next
        
        # Decrease the length by 1
        self.length -= 1
        
        if self.length == 0:
            self.tail = None
            
        # Set the next of old_head to None
        # and return the old_head
        old_head.next = None
        
        return old_head
            
    # Unshift head
    # Create new node and set that as head
    def unshift(self, data):
        # Create a new node from provided data
        new_head = Node(data)
        
        if self.head is None:
            self.tail = new_head
        else:
            new_head.next = self.head
            
        # Set the new_head as head of the list
        self.head = new_head
        
        # Increase size of the list by 1
        self.length += 1
        
        return new_head
        
    # Get node by index(sequence numbers)
    # 0 based index (index starts from 0)
    def get(self, index):
        # Check if provided inex is valid
        # Return None if index is invalid
        if index < 0 or index >= self.length:
            return None
        
        selected_node = self.head
        
        # Start from head and go till the provided index
        for _ in range(index):
            selected_node = selected_node.next
            
        return selected_node
    
    # Set data of a specific node at specific index
    # 0 based index(index starts from 0)
    def set(self, index, data):
        node_at_index = self.get(index)
        
        # If item exists at index then return that
        if node_at_index is not None:
            node_at_index.data = data
            
            return True
        
        return False
    
    # Insert new node at a specific index
    # 0 based index(index starts at 0)
    def insert(self, index, data):
        # Return false if index is out of range
        if index < 0 or index > self.length:
            return False
        
        # If index is the next element after tail
        # then just push item to the lsit
        if index == self.length:
            self.push(data)
            
            return True
        
        # If index is zero then unshift
        if index == 0:
            self.unshift(data)
            
            return True
        
        # Create new node and add it to the defined index
        new_node = Node(data)
        
        # Get the node after which this new node will be inserted
        prev_node = self.get(index - 1)
        
        # Add the new_node after the prev_node
        new_node.next = prev_node.next
        prev_node.next = new_node
        
        # Increase length by one
        self.length += 1
        
        return True
          
    # Reemove item at specific index
    def remove(self, index):
        # Return false if index is out of range
        if index < 0 or index >= self.length:
            return None
        
        if index == 0:
            return self.shift()
        
        if index == self.length - 1:
            return self.pop()
        
        prev_node = self.get(index - 1)
        node_to_remove = prev_node.next
        prev_node.next = prev_node.next.next
        
        # Decrease length by 1
        self.length -= 1
        
        return node_to_remove 
        
    # Reverse a linked list
    def reverse(self):
        current_node = self.head
        prev_node = None
        next_node = None
        
        self.tail = self.head
        
        while current_node:
            next_node = current_node.next
            current_node.next = prev_node

            prev_node = current_node
            current_node = next_node

        self.head = prev_node
        
        return self
    
    # Search the list for specific data
    def search(self, data):
        current = self.head
        index = 0
        
        while current:
            if current.data == data:
                return index
            
            current = current.next
            index += 1
            
        return None
        
    
    def traverse(self):
        current = self.head
        
        while current:
            print(current.data, end=" -> ")
            
            # Move to the next node
            current = current.next
            